zone/raids.cpp

```diff
@@ -2203,3 +2257,734 @@ void Raid::SetNewRaidLeader(uint32 i)
+	DelegateAbility_Struct* das = (DelegateAbility_Struct*)outapp->pBuffer;
+	if (mm) {
+		das->Action = ClearDelegate;
+		memset(main_marker_pcs[mm - 1], 0, 64);
+		rm->main_marker = 0;
+		std::string query = StringFormat("UPDATE raid_members SET is_marker = 0 WHERE raidid = %i AND name= '%s';",
+			GetID(),
+			delegatee
+		);
+		auto results = database.QueryDatabase(query);
+		if (!results.Success()) {
+			LogError("Unable to clear raid main marker for player: [{}] with error [{}]\n",
+				delegatee,
+				results.ErrorMessage().c_str()
+			);
+		}
+	}
+	else {
+		if (slot >= 0) {
+			strcpy(main_marker_pcs[slot], c->GetName());
+			rm->main_marker = slot + 1;
+			das->Action = SetDelegate;
+			std::string query = StringFormat("UPDATE raid_members SET is_marker = %i WHERE raidid = %i AND name= '%s';",
+				slot + 1,
+				this->GetID(),
+				delegatee
+			);
+			auto results = database.QueryDatabase(query);
+			if (!results.Success()) {
+				LogError("Unable to set raid main marker for player: [{}] slot: [{}] with error [{}]\n",
+					delegatee,
+					slot + 1,
+					results.ErrorMessage().c_str()
+				);
+			}
+		}
+	}
+	das->DelegateAbility = RaidDelegateMainMarker;
+	das->MemberNumber = 0;
+	das->EntityID = c->GetID();
+	strcpy(das->Name, delegatee);
+	QueuePacket(outapp);
+	safe_delete(outapp);
+}
+
+int Raid::FindNextRaidDelegateSlot(int option) 
+{
+	//option 1 FindNextRaidMainMarkerSlot
+	//option 2 FindNextRaidMainAssisterSlot
+	if (option == 1) {
+		for (int i = 0; i < MAX_NO_RAID_MAIN_MARKERS; i++) {
+			if (strlen(main_marker_pcs[i]) == 0) {
+					return i;
+			}
+		}
+		return -1;		
+	}
+	else if (option == 2) {
+		for (int i = 0; i < MAX_NO_RAID_MAIN_ASSISTERS; i++) {
+			if (strlen(main_assister_pcs[i]) == 0) {
+				return i;
+			}
+		}
+		return -1;		
+	}
+}
+
+void Raid::UpdateXTargetType(XTargetType Type, Mob* m, const char* Name)
+{
+	for (const auto& rm : members) {
+		if (!rm.member || rm.is_bot || !rm.member->XTargettingAvailable())
+			continue;
+
+		for (int i = 0; i < rm.member->GetMaxXTargets(); ++i)
+		{
+			if (rm.member->XTargets[i].Type == Type)
+			{
+				if (m) {
+					rm.member->XTargets[i].ID = m->GetID();
+				}
+				else {
+					rm.member->XTargets[i].ID = 0;
+				}
+
+				if (Name) {
+					strncpy(rm.member->XTargets[i].Name, Name, 64);
+				}
+
+				rm.member->SendXTargetPacket(i, m);
+			}
+		}
+	}
+}
+
+void Raid::RaidMarkNPC(Mob* mob, uint32 parameter)
+{
+	Client* c = mob->CastToClient();
+	if (!c || !c->GetTarget() || parameter < 1 || parameter > 3) {
+		LogDebug("RaidMarkNPC Failed sanity checks.");
+		return;
+	}
+
+	for (int i = 0; i < MAX_NO_RAID_MAIN_MARKERS; i++) {
+		auto cname = c->GetCleanName();
+		if (strcasecmp(main_marker_pcs[i], cname) == 0 || strcasecmp(leadername, cname) == 0) {
+			marked_npcs[parameter - 1] = c->GetTarget()->GetID();
+
+			std::string query = StringFormat("UPDATE raid_details SET marked_npc%i = %i WHERE raidid = %i;",
```

Comment:
Can make a repository method for this in `common/repositories/raid_details_repository.h` as well.
```cpp
static int UpdateRaidMarkedNPC(
	Database& db,
	int32_t raid_id,
    uint8_t marked_index,
    uint16_t marked_npc_id
) {
	auto results = db.QueryDatabase(
		fmt::format(
			"UPDATE {} SET marked_npc{} = {} WHERE raidid = {}",
			TableName(),
			marked_index,
			marked_npc_id,
			raid_id
		)
	);

	return results.Success() ? results.RowsAffected() : 0;
}
```

Then just do this below.
```cpp
int updated = RaidDetailsRepository::UpdateRaidMarkedNPC(database, GetID(), parameter, marked_npcs[parameter - 1]);
if (!updated) {
    LogError("Unable to set MarkedNPC{} from slot [{}]", parameter, parameter - 1);
    return;
}
```

Reply:
Updated.